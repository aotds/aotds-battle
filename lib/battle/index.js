import { createStore, applyMiddleware } from "redux";
import _ from 'lodash';

import battle_reducer from './Reducer';
import { validate } from './schema';

import logger from '../Logger';

import Action from './Actions';

/**
 * build middlewares for only certain types of actions
 * @param actions list of action types the middleware should be invoked for
 */
function for_actions(...actions) {
    return store => next => action => {
        if (  actions.indexOf( action.type ) > -1 ) {
            this(store)(next)(action);
        }
        else {
            next(action);
        }
    };
}

const MW_firecons_fire = ( store => next => action => {
    let state = store.getState();

    next(action);

    _( _(state.objects).find({ id: action.object_id }) )
        .get( 'firecons', [] )
        .filter( f => f.target_id )
        .forEach( f => {
            let a = Action.firecon_fire({ object_id: action.object_id, firecon_id: f.id });
            console.log( 'a', a );
            store.dispatch(a);
        });
} ) ::for_actions( Action.FIRECONS_FIRE );

const MW_firecon_fire = ( store => next => action => {
    let state = store.getState();

    next(action);

    let ship = _.find( state.objects, { id: action.object_id } );

    let firecon = _.find( ship.firecons, { id: action.firecon_id } );

     _.get( ship, 'weapons', [] )
        .filter( f => f.firecon_id === firecon.id )
        .forEach( f => {
            let a = Action.weapon_fire({ object_id: action.object_id, weapon_id: f.id, target_id: firecon.target_id });
            store.dispatch(a);
        });
} ) ::for_actions( Action.FIRECON_FIRE );

function percolate_action() {
    return  store => next => action => {
        next(action);
        this(store)(next)(action);
    };
}

const find_object  = function(id) { return _.find( this.objects,  { id } ) };
const find_firecon = function(id) { return _.find( this.firecons, { id } ) };
const find_weapon  = function(id) { return _.find( this.weapons,  { id } ) };


let middlewares = [
    MW_firecons_fire,
    MW_firecon_fire,
];

import createSagaMiddleware from 'redux-saga';
import battle_saga from './saga';


export default
class Battle {
    store;

    constructor ( options = {} ) {
        const sagaMiddleware = createSagaMiddleware();

        this.store = createStore( 
            battle_reducer,
            options.state || {},
            applyMiddleware( ...middlewares, sagaMiddleware ),
        );
        sagaMiddleware.run( battle_saga );

        // TODO only check for the schema in dev mode
        this.store.subscribe( () => {
            this.store.getState()::validate( 'battle' );
        })
    }

    init_game = payload => this.store.dispatch(
        Action.init_game(payload)
    ) 

    get state () { return this.store.getState() }

    dispatch( action ) {
        return this.store.dispatch(action);
    }

};


