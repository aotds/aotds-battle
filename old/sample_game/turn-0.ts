export const actions = [];

export const dice = [];

export const tests = (state) => {
	expect(state).toMatchObject({});
};
