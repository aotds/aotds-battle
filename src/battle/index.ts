import fp from 'lodash/fp';

import { compose, createStore, applyMiddleware } from "redux";
import { Action } from '../reducer/types';
import reducer from '../store/reducer';
import { try_play_turn } from '../store/actions/phases';
import { log_skipper } from '../store/log/middleware';
import { timestamp } from '../store/middleware/timestamp';
import { action_id_mw_gen } from '../store/middleware/action_id';
import mw_play_phases from '../store/middleware/play_phases';
import mw_bogeys_movement from '../middleware/movement_phase';
import mw_bogey_firecon_orders from '../middleware/firecons_phase';
import mw_bogey_weapon_orders from '../middleware/weapons_phase';

type BattleOpts = {
    devtools?: {},
    state?: {}
}

export default class Battle {

    store: any;

    constructor( opts: BattleOpts ) {

        let enhancers = applyMiddleware(
            log_skipper([ 'TRY_PLAY_TURN' ]),
            timestamp,
            action_id_mw_gen(),
            mw_play_phases,
            mw_bogeys_movement,
            mw_bogey_firecon_orders,
            mw_bogey_weapon_orders,
        );

        if( opts.devtools) {
            enhancers = require('remote-redux-devtools').composeWithDevTools(
                fp.defaults({port: 8000, hostname: 'localhost'}, opts.devtools)
            )( enhancers )
        }

        this.store = createStore(
            reducer,
            opts.state,
            enhancers,
        );
    }

    get state() { return this.store.getState() }

    dispatch(action :Action) {
        this.store.dispatch(action);
    }

}
