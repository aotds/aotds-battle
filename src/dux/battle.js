import { BDux } from '../BDux.js';

import { dux as actionId, addActionIdEffect } from './actionId.js';
import { dux as log } from './log.js';

export const battleDux = new BDux({
	subduxes: {
		actionId,
		log,
	},
});

battleDux.addEffect('*', addActionIdEffect);
