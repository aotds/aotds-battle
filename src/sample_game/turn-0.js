import { expect, test } from 'vitest';

export const tests = (state) => {
	test('we have our subduxes', () => {
		expect(state).toHaveProperty('actionId');
		expect(state).toHaveProperty('log');
	});
};
