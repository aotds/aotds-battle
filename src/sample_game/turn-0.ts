export const actions = [];

export const dice = [];

export const tests = state => t => {
    // nothing to see here
    expect(true).toBeTruthy();
};
