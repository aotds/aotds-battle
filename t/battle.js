import tap from 'tap';
import _ from 'lodash';

import Battle from '../lib/battle';

import logger from '../lib/Logger';
logger.level = 'debug';

tap.test( 'create battle', tap => {

    const battle = new Battle();

    battle.init_game( {
        name: 'gemini',
        objects: [
            { name: 'Enkidu' },
            { name: 'Siduri' },
        ],
    });

    let state = battle.state;

    tap.same( state.game, { name: 'gemini' }, 'game' );
    tap.same( state.objects, [ 
        { name: 'Enkidu' }, 
        { name: 'Siduri' } 
    ], 'objects' );

    tap.end();
})

