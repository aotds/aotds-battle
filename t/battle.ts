import tap = require('tap');
import _ from 'lodash';

import Battle from '../lib/battle';

interface Logger {
    level?: string;
    debug?: any;
}

import logger from '../lib/Logger';
(<Logger>logger).level = 'debug';

tap.test( 'create battle', tap => {

    const battle = new Battle();

    battle.init_game( {
        name: 'gemini',
        objects: [
            { name: 'Enkidu' },
            { name: 'Siduri' },
        ],
    });

    let state = battle.state;

    tap.same( state.game, { name: 'gemini' }, 'game' );
    tap.same( state.objects, [ 
        { name: 'Enkidu' }, 
        { name: 'Siduri' } 
    ], 'objects' );

    tap.end();
})

import Schema from '../lib/battle/Schema';

console.log( JSON.stringify( Schema ) )
