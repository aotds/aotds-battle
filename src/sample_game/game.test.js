import { test, describe, expect } from 'vitest';

import { battleDux } from '../dux/battle.js';

import * as Turn0 from './turn-0.js';
//import * as Turn1 from './turn-1.js';

const battle = battleDux.createStore();

describe.each([Turn0])('turn %#', async ({ actions = [], tests }) => {
	actions.forEach(battle.dispatch);

	const state = battle.getState();

	if (tests) describe('tests', () => tests(state, battle));

	test('snapshot', () => expect(state).toMatchSnapshot());
});
