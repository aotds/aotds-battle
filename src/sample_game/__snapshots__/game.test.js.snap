// Vitest Snapshot v1

exports[`turn 0 > snapshot 1`] = `
{
  "actionId": 1,
  "log": [],
}
`;
